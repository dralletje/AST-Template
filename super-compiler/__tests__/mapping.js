let { create_context } = require("../mapping.js");

it("should do let_assigment", () => {
  expect(
    create_context().evaluate_statements(`let x = "hey";`)
  ).toMatchSnapshot();
});

it("should do if_statement with statements", () => {
  expect(
    create_context().evaluate_statements(`
      if (3 > x) {
        console.log('x');
      }
    `)
  ).toMatchSnapshot();
});

it("should do if_else with statements", () => {
  expect(
    create_context().evaluate_statements(`
      if (x > 3) {
        console.log('x');
      } else {
        console.log('y');
      }
    `)
  ).toMatchSnapshot();
});

it("should do gte_operator", () => {
  expect(
    create_context().evaluate_expression(`10 > 4`)
  ).toMatchSnapshot();
});

it("should do function_call", () => {
  expect(
    create_context().evaluate_expression(`func()`)
  ).toMatchSnapshot();
});
it("should do function_call with arguments", () => {
  expect(
    create_context().evaluate_expression(`func(10, "hey", func())`)
  ).toMatchSnapshot();
});
it("should do function_call with arguments", () => {
  expect(
    create_context().evaluate_expression(`func(10, "hey", func())`)
  ).toMatchSnapshot();
});
it("should do variable", () => {
  expect(
    create_context().evaluate_expression(`func`)
  ).toMatchSnapshot();
});
it("should do return_statement", () => {
  expect(
    create_context().evaluate_statements(`return x`)
  ).toMatchSnapshot();
});

it("should do object_property", () => {
  expect(
    create_context().evaluate_statements(`object_variable.property_name`)
  ).toMatchSnapshot();
});

it("should do constant_literal string", () => {
  expect(
    create_context().evaluate_expression(`"Hey"`)
  ).toMatchSnapshot();
});
it("should do constant_literal number", () => {
  expect(
    create_context().evaluate_statements(`10`)
  ).toMatchSnapshot();
});

it("should parse program", () => {
  let simple_program = `
    let x = Math.random();
    if (x > 2) {
      return "Universe broke";
    } else {
      return "Good";
    }
  `;

  let context = create_context();

  let assertions = context.evaluate_statements(simple_program);

  expect(assertions).toMatchSnapshot();
});
