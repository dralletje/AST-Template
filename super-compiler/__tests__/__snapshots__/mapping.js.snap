// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should do constant_literal number 1`] = `
Array [
  Array [
    Object {
      "argument_one": Symbol(Allocation @ root.1.1),
      "argument_two": 10,
      "operator": "creates",
      "type": "Assert",
    },
    Object {
      "argument_one": Symbol(Return reference @ root.1),
      "argument_two": Symbol(Allocation @ root.1.1),
      "operator": "becomes",
      "type": "Assert",
    },
  ],
]
`;

exports[`should do constant_literal string 1`] = `
Object {
  "assertions": Array [
    Object {
      "argument_one": Symbol(Allocation @ root.1.1),
      "argument_two": "Hey",
      "operator": "creates",
      "type": "Assert",
    },
    Object {
      "argument_one": Symbol(Return reference @ root.1),
      "argument_two": Symbol(Allocation @ root.1.1),
      "operator": "becomes",
      "type": "Assert",
    },
  ],
  "return": Symbol(Return reference @ root.1),
}
`;

exports[`should do function_call 1`] = `
Object {
  "assertions": Array [
    Array [
      Object {
        "argument_one": Symbol(Return reference @ root.1.1),
        "argument_two": Object {
          "argument_one": Symbol(Current scope reference @ root.1.1),
          "argument_two": "func",
          "operator": ".",
          "type": "Express",
        },
        "operator": "becomes",
        "type": "Assert",
      },
    ],
    Array [],
    Object {
      "argument_one": Symbol(Return reference @ root.1.1),
      "argument_two": Array [],
      "operator": "()",
      "type": "Assert",
    },
  ],
  "return": Symbol(Return reference @ root.1),
}
`;

exports[`should do function_call with arguments 1`] = `
Object {
  "assertions": Array [
    Array [
      Object {
        "argument_one": Symbol(Return reference @ root.1.1),
        "argument_two": Object {
          "argument_one": Symbol(Current scope reference @ root.1.1),
          "argument_two": "func",
          "operator": ".",
          "type": "Express",
        },
        "operator": "becomes",
        "type": "Assert",
      },
    ],
    Array [
      Array [
        Object {
          "argument_one": Symbol(Allocation @ root.1.2.1),
          "argument_two": 10,
          "operator": "creates",
          "type": "Assert",
        },
        Object {
          "argument_one": Symbol(Return reference @ root.1.2),
          "argument_two": Symbol(Allocation @ root.1.2.1),
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Array [
        Object {
          "argument_one": Symbol(Allocation @ root.1.3.1),
          "argument_two": "hey",
          "operator": "creates",
          "type": "Assert",
        },
        Object {
          "argument_one": Symbol(Return reference @ root.1.3),
          "argument_two": Symbol(Allocation @ root.1.3.1),
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Array [
        Array [
          Object {
            "argument_one": Symbol(Return reference @ root.1.4.1),
            "argument_two": Object {
              "argument_one": Symbol(Current scope reference @ root.1.4.1),
              "argument_two": "func",
              "operator": ".",
              "type": "Express",
            },
            "operator": "becomes",
            "type": "Assert",
          },
        ],
        Array [],
        Object {
          "argument_one": Symbol(Return reference @ root.1.4.1),
          "argument_two": Array [],
          "operator": "()",
          "type": "Assert",
        },
      ],
    ],
    Object {
      "argument_one": Symbol(Return reference @ root.1.1),
      "argument_two": Array [
        undefined,
        undefined,
        undefined,
      ],
      "operator": "()",
      "type": "Assert",
    },
  ],
  "return": Symbol(Return reference @ root.1),
}
`;

exports[`should do function_call with arguments 2`] = `
Object {
  "assertions": Array [
    Array [
      Object {
        "argument_one": Symbol(Return reference @ root.1.1),
        "argument_two": Object {
          "argument_one": Symbol(Current scope reference @ root.1.1),
          "argument_two": "func",
          "operator": ".",
          "type": "Express",
        },
        "operator": "becomes",
        "type": "Assert",
      },
    ],
    Array [
      Array [
        Object {
          "argument_one": Symbol(Allocation @ root.1.2.1),
          "argument_two": 10,
          "operator": "creates",
          "type": "Assert",
        },
        Object {
          "argument_one": Symbol(Return reference @ root.1.2),
          "argument_two": Symbol(Allocation @ root.1.2.1),
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Array [
        Object {
          "argument_one": Symbol(Allocation @ root.1.3.1),
          "argument_two": "hey",
          "operator": "creates",
          "type": "Assert",
        },
        Object {
          "argument_one": Symbol(Return reference @ root.1.3),
          "argument_two": Symbol(Allocation @ root.1.3.1),
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Array [
        Array [
          Object {
            "argument_one": Symbol(Return reference @ root.1.4.1),
            "argument_two": Object {
              "argument_one": Symbol(Current scope reference @ root.1.4.1),
              "argument_two": "func",
              "operator": ".",
              "type": "Express",
            },
            "operator": "becomes",
            "type": "Assert",
          },
        ],
        Array [],
        Object {
          "argument_one": Symbol(Return reference @ root.1.4.1),
          "argument_two": Array [],
          "operator": "()",
          "type": "Assert",
        },
      ],
    ],
    Object {
      "argument_one": Symbol(Return reference @ root.1.1),
      "argument_two": Array [
        undefined,
        undefined,
        undefined,
      ],
      "operator": "()",
      "type": "Assert",
    },
  ],
  "return": Symbol(Return reference @ root.1),
}
`;

exports[`should do gte_operator 1`] = `
Object {
  "assertions": Array [
    Array [
      Object {
        "argument_one": Symbol(Allocation @ root.1.1.1),
        "argument_two": 10,
        "operator": "creates",
        "type": "Assert",
      },
      Object {
        "argument_one": Symbol(Return reference @ root.1.1),
        "argument_two": Symbol(Allocation @ root.1.1.1),
        "operator": "becomes",
        "type": "Assert",
      },
    ],
    Array [
      Object {
        "argument_one": Symbol(Allocation @ root.1.2.1),
        "argument_two": 4,
        "operator": "creates",
        "type": "Assert",
      },
      Object {
        "argument_one": Symbol(Return reference @ root.1.2),
        "argument_two": Symbol(Allocation @ root.1.2.1),
        "operator": "becomes",
        "type": "Assert",
      },
    ],
    Object {
      "condition": Object {
        "argument_one": Symbol(Return reference @ root.1.1),
        "argument_two": Symbol(Return reference @ root.1.2),
        "operator": ">",
        "type": "Assert",
      },
      "inverse": false,
      "type": "Condition",
    },
  ],
  "return": Symbol(Return reference @ root.1),
}
`;

exports[`should do if_else with statements 1`] = `
Array [
  Array [
    Array [
      Array [
        Object {
          "argument_one": Symbol(Return reference @ root.8.1.1),
          "argument_two": Object {
            "argument_one": Symbol(Current scope reference @ root.8.1.1),
            "argument_two": "x",
            "operator": ".",
            "type": "Express",
          },
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Array [
        Object {
          "argument_one": Symbol(Allocation @ root.8.1.2.1),
          "argument_two": 3,
          "operator": "creates",
          "type": "Assert",
        },
        Object {
          "argument_one": Symbol(Return reference @ root.8.1.2),
          "argument_two": Symbol(Allocation @ root.8.1.2.1),
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Object {
        "condition": Object {
          "argument_one": Symbol(Return reference @ root.8.1.1),
          "argument_two": Symbol(Return reference @ root.8.1.2),
          "operator": ">",
          "type": "Assert",
        },
        "inverse": false,
        "type": "Condition",
      },
    ],
    Object {
      "assertions": undefined,
      "condition": Object {
        "condition": undefined,
        "inverse": false,
        "type": "Condition",
      },
      "type": "Fork",
    },
    Object {
      "assertions": undefined,
      "condition": Object {
        "condition": undefined,
        "inverse": true,
        "type": "Condition",
      },
      "type": "Fork",
    },
  ],
]
`;

exports[`should do if_statement with statements 1`] = `
Array [
  Array [
    Array [
      Array [
        Object {
          "argument_one": Symbol(Allocation @ root.7.1.1.1),
          "argument_two": 3,
          "operator": "creates",
          "type": "Assert",
        },
        Object {
          "argument_one": Symbol(Return reference @ root.7.1.1),
          "argument_two": Symbol(Allocation @ root.7.1.1.1),
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Array [
        Object {
          "argument_one": Symbol(Return reference @ root.7.1.2),
          "argument_two": Object {
            "argument_one": Symbol(Current scope reference @ root.7.1.2),
            "argument_two": "x",
            "operator": ".",
            "type": "Express",
          },
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Object {
        "condition": Object {
          "argument_one": Symbol(Return reference @ root.7.1.1),
          "argument_two": Symbol(Return reference @ root.7.1.2),
          "operator": ">",
          "type": "Assert",
        },
        "inverse": false,
        "type": "Condition",
      },
    ],
    Object {
      "assertions": undefined,
      "condition": Object {
        "condition": undefined,
        "inverse": false,
        "type": "Condition",
      },
      "type": "Fork",
    },
  ],
]
`;

exports[`should do let_assigment 1`] = `
Array [
  Array [
    Array [
      Object {
        "argument_one": Symbol(Allocation @ root.6.1.1),
        "argument_two": "hey",
        "operator": "creates",
        "type": "Assert",
      },
      Object {
        "argument_one": Symbol(Return reference @ root.6.1),
        "argument_two": Symbol(Allocation @ root.6.1.1),
        "operator": "becomes",
        "type": "Assert",
      },
    ],
    Object {
      "argument_one": Object {
        "argument_one": Symbol(Current scope reference @ root.6),
        "argument_two": "x",
        "operator": ".",
        "type": "Express",
      },
      "argument_two": Symbol(Return reference @ root.6.1),
      "operator": "becomes",
      "type": "Assert",
    },
  ],
]
`;

exports[`should do object_property 1`] = `
Array [
  Array [
    Array [
      Object {
        "argument_one": Symbol(Return reference @ root.5.1),
        "argument_two": Object {
          "argument_one": Symbol(Current scope reference @ root.5.1),
          "argument_two": "object_variable",
          "operator": ".",
          "type": "Express",
        },
        "operator": "becomes",
        "type": "Assert",
      },
    ],
    Array [
      Object {
        "argument_one": Symbol(Allocation @ root.5.2.1),
        "argument_two": "property_name",
        "operator": "creates",
        "type": "Assert",
      },
      Object {
        "argument_one": Symbol(Return reference @ root.5.2),
        "argument_two": Symbol(Allocation @ root.5.2.1),
        "operator": "becomes",
        "type": "Assert",
      },
    ],
    Object {
      "argument_one": Symbol(Return reference @ root.5),
      "argument_two": Object {
        "argument_one": Symbol(Return reference @ root.5.1),
        "argument_two": Symbol(Return reference @ root.5.2),
        "operator": ".",
        "type": "Express",
      },
      "operator": "becomes",
      "type": "Assert",
    },
  ],
]
`;

exports[`should do return_statement 1`] = `
Array [
  Array [
    Array [
      Object {
        "argument_one": Symbol(Return reference @ root.9.1),
        "argument_two": Object {
          "argument_one": Symbol(Current scope reference @ root.9.1),
          "argument_two": "x",
          "operator": ".",
          "type": "Express",
        },
        "operator": "becomes",
        "type": "Assert",
      },
    ],
    Object {
      "to": Symbol(Jump return @root),
      "type": "Jump",
      "with": Symbol(Return reference @ root.9.1),
    },
  ],
]
`;

exports[`should do variable 1`] = `
Object {
  "assertions": Array [
    Object {
      "argument_one": Symbol(Return reference @ root.1),
      "argument_two": Object {
        "argument_one": Symbol(Current scope reference @ root.1),
        "argument_two": "func",
        "operator": ".",
        "type": "Express",
      },
      "operator": "becomes",
      "type": "Assert",
    },
  ],
  "return": Symbol(Return reference @ root.1),
}
`;

exports[`should parse program 1`] = `
Array [
  Array [
    Array [
      Array [
        Array [
          Object {
            "argument_one": Symbol(Return reference @ root.6.1.1.1),
            "argument_two": Object {
              "argument_one": Symbol(Current scope reference @ root.6.1.1.1),
              "argument_two": "Math",
              "operator": ".",
              "type": "Express",
            },
            "operator": "becomes",
            "type": "Assert",
          },
        ],
        Array [
          Object {
            "argument_one": Symbol(Allocation @ root.6.1.1.2.1),
            "argument_two": "random",
            "operator": "creates",
            "type": "Assert",
          },
          Object {
            "argument_one": Symbol(Return reference @ root.6.1.1.2),
            "argument_two": Symbol(Allocation @ root.6.1.1.2.1),
            "operator": "becomes",
            "type": "Assert",
          },
        ],
        Object {
          "argument_one": Symbol(Return reference @ root.6.1.1),
          "argument_two": Object {
            "argument_one": Symbol(Return reference @ root.6.1.1.1),
            "argument_two": Symbol(Return reference @ root.6.1.1.2),
            "operator": ".",
            "type": "Express",
          },
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Array [],
      Object {
        "argument_one": Symbol(Return reference @ root.6.1.1),
        "argument_two": Array [],
        "operator": "()",
        "type": "Assert",
      },
    ],
    Object {
      "argument_one": Object {
        "argument_one": Symbol(Current scope reference @ root.6),
        "argument_two": "x",
        "operator": ".",
        "type": "Express",
      },
      "argument_two": Symbol(Return reference @ root.6.1),
      "operator": "becomes",
      "type": "Assert",
    },
  ],
  Array [
    Array [
      Array [
        Object {
          "argument_one": Symbol(Return reference @ root.14.1.1),
          "argument_two": Object {
            "argument_one": Symbol(Current scope reference @ root.14.1.1),
            "argument_two": "x",
            "operator": ".",
            "type": "Express",
          },
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Array [
        Object {
          "argument_one": Symbol(Allocation @ root.14.1.2.1),
          "argument_two": 2,
          "operator": "creates",
          "type": "Assert",
        },
        Object {
          "argument_one": Symbol(Return reference @ root.14.1.2),
          "argument_two": Symbol(Allocation @ root.14.1.2.1),
          "operator": "becomes",
          "type": "Assert",
        },
      ],
      Object {
        "condition": Object {
          "argument_one": Symbol(Return reference @ root.14.1.1),
          "argument_two": Symbol(Return reference @ root.14.1.2),
          "operator": ">",
          "type": "Assert",
        },
        "inverse": false,
        "type": "Condition",
      },
    ],
    Object {
      "assertions": undefined,
      "condition": Object {
        "condition": undefined,
        "inverse": false,
        "type": "Condition",
      },
      "type": "Fork",
    },
    Object {
      "assertions": undefined,
      "condition": Object {
        "condition": undefined,
        "inverse": true,
        "type": "Condition",
      },
      "type": "Fork",
    },
  ],
]
`;
